### ISSUE TITLE
Puma Cluster Workers Sleeping and unkillable using SIGTERM

### ISSUE THREAD

**Issue by straleyb** [puma/puma#1926]
### Steps to reproduce

Im not entirely sure how to reproduce this error since it only is happening on our production server. However, I'll try to express the issue as best as I can in hopes that either someone has a solution or we uncovered a bug of some kind.

The problem is this. As noted many times, web servers begin to use up more and more memory as time goes on. After all the memory on the system gets eaten up, the system crashes due to not enough RAM. This is why PWK was introduced. PWK can kill these Puma Cluster Workers, which are eating up the memory and free up that space. Then, a new worker can by spun up and all is well. 

On our production server, however, we are not able to kill these Puma Cluster Workers. They are entering S mode (which inhibits the ability to receive SIGTERM signals) and will not be killed. But the code still allows for the creation of another Puma Cluster Worker in its place. 

This means that now there are 4 Puma Cluster Workers with a parent and 1 Puma Cluster Worker that is orphaned and blocking. This happens on a continual basis and we end up with 15 to 20 Puma Cluster Workers over the course of several hours, that are all sucking up our memory until its completely used up.

We were able to write a reaper which calls SIGKILL on the orphaned processes and removes them (this is done with some trickery using pumactl stats and diffing them against the current list of processes found using `ps aux | grep puma`. Then once we find the diff, we kill the orphaned processes.

This is just a bandaid to the issue though. We would like to find an actual fix.

Im not sure if this is related to #1855 since the workers arent marked "defunct" so I figured id make a new ticket instead. Any help would be greatly appreciated and ill answer any other questions that pop up.

```
ps -ef | grep puma
deploy-+  9572  8193  0 09:43 pts/3    00:00:07 watch /usr/local/sbin/watch-puma.sh
deploy-+ 10322 63173 17 13:05 ?        00:03:38 puma: cluster worker 0: 63173 [20190820172048]
deploy-+ 10325 63173 16 13:05 ?        00:03:35 puma: cluster worker 1: 63173 [20190820172048]
deploy-+ 10328 63173 18 13:05 ?        00:03:53 puma: cluster worker 2: 63173 [20190820172048]
deploy-+ 10331 63173 17 13:05 ?        00:03:44 puma: cluster worker 3: 63173 [20190820172048]
deploy-+ 34188  2897  0 13:26 pts/0    00:00:00 grep --color=auto puma
deploy-+ 63173     1  0 09:42 ?        00:00:08 puma 3.12.1 (unix:///data0/hydra/shared/sockets/puma.sock) [20190820172048]
[deploy-rails@lib-saapp1 ~]$ kill 10325
[deploy-rails@lib-saapp1 ~]$ ps -ef | grep puma
deploy-+  9572  8193  0 09:43 pts/3    00:00:07 watch /usr/local/sbin/watch-puma.sh
deploy-+ 10322 63173 17 13:05 ?        00:03:39 puma: cluster worker 0: 63173 [20190820172048]
deploy-+ 10325 63173 16 13:05 ?        00:03:35 puma: cluster worker 1: 63173 [20190820172048]
deploy-+ 10328 63173 18 13:05 ?        00:03:54 puma: cluster worker 2: 63173 [20190820172048]
deploy-+ 10331 63173 17 13:05 ?        00:03:46 puma: cluster worker 3: 63173 [20190820172048]
deploy-+ 35067 63173  2 13:26 ?        00:00:00 puma: cluster worker 1: 63173 [20190820172048]
deploy-+ 35371  2897  0 13:26 pts/0    00:00:00 grep --color=auto puma
deploy-+ 63173     1  0 09:42 ?        00:00:08 puma 3.12.1 (unix:///data0/hydra/shared/sockets/puma.sock) [20190820172048]
```
That is the output of what happens when I try to kill an orphaned process and it doesnt actually die....

### System configuration

**Ruby version**: Ruby 2.5.5
**Rails version**: Rails 5.0.7.2
**Puma version**: puma (3.12.1)

**Comment by MSP-Greg** [comment]
The issue of 'nonresponsive' workers is messy, but there have been recent commits attempting to address the issue.  If you're able to, testing with current master may have different results.

Note that Puma defaults to allowing 'nonresponsive' workers 30 seconds or so before it changes from SIGTERM to SIGKILL.  After that timeout, it should kill the worker and replace it.

**Comment by nateberkopec** [comment]
Yeah, it looks like you waited about a second after SIGTERM before checking the process table again. Please try waiting 30-60 seconds and see if Puma's timeout works.

**Comment by nateberkopec** [comment]
Good report but we'll definitely need a minimal reproducing app that doesn't exit properly after ~60 seconds.

> As noted many times, web servers begin to use up more and more memory as time goes on.

Unrelated, but you should be [trying jemalloc or MALLOC_ARENA_MAX.](https://www.speedshop.co/2017/12/04/malloc-doubles-ruby-memory.html)

> They are entering S mode (which inhibits the ability to receive SIGTERM signals) and will not be killed

? S is interruptible sleep, it doesnt' change how puma processes signals, unless you know something I don't (which you may!).

> Im not sure if this is related to #1855

Correct, it isn't. That was for defunct table entries, different issue.

**Comment by nateberkopec** [comment]
@straleyb Did you reproduce the issue on master?

**Comment by nateberkopec** [comment]
Closing. If you can create a reproducible example, please open a new issue.

**Comment by sanjayprabhu** [comment]
We ran into the same issue.

Here's what we noticed:
 - Puma cluster master process would not respond to sigterms, and neither did the workers
 - A subset of workers (let's call them idle workers) were not actually processing requests
 - There were a lot of "Terminating timed out worker .." in our logs on deploy.
 - Calling `pumactl stats` showed that the master thought there were 0 booted workers (even though some were serving traffic)

After debugging for a while, it appears that the master process and the idle workers were both blocked on reading/writing to the worker-master pipe. I'm not sure how they get into this state, but seems like it happens when the workers take a lot of time to boot up, and the master has to terminated many of them. 

We worked around this by increasing the `worker_boot_timeout`. But it would be great to fix the underlying issue to make sure it can't get into this state. We're on puma `3.12.2`, however, I noticed the same issue on master.

Here's are the Thread backtraces from the different kinds of workers to show what's going on:

Puma cluster master process
```
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:238:in `write'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:238:in `wakeup!'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:370:in `block in setup_signals'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:201:in `sleep'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:201:in `check_workers'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:486:in `run'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/launcher.rb:186:in `run'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cli.rb:80:in `run'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/bin/puma:10:in `<top (required)>'
/home/app/vendor/bundle/ruby/2.6.0/bin/puma:23:in `load'
/home/app/vendor/bundle/ruby/2.6.0/bin/puma:23:in `<top (required)>'
/usr/local/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/cli/exec.rb:74:in `load'
/usr/local/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/cli/exec.rb:74:in `kernel_load'
/usr/local/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/cli/exec.rb:28:in `run'
/usr/local/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/cli.rb:463:in `exec'
```

Puma "idle" worker
```
# It did not have a puma server loop thread, because it hasn't gotten to the point where it starts the server

/home/app/lib/puma_helpers/hooks.rb:55:in `backtrace'
/home/app/lib/puma_helpers/hooks.rb:55:in `block (2 levels) in after_worker_fork'
/home/app/lib/puma_helpers/hooks.rb:47:in `each'
/home/app/lib/puma_helpers/hooks.rb:47:in `block in after_worker_fork'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:282:in `write'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:282:in `<<'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:282:in `worker'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:139:in `block (2 levels) in spawn_workers'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:139:in `fork'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:139:in `block in spawn_workers'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:135:in `times'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:135:in `spawn_workers'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:213:in `check_workers'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:486:in `run'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/launcher.rb:186:in `run'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cli.rb:80:in `run'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/bin/puma:10:in `<top (required)>'
/home/app/vendor/bundle/ruby/2.6.0/bin/puma:23:in `load'
/home/app/vendor/bundle/ruby/2.6.0/bin/puma:23:in `<top (required)>'

```
Here's the strace of when we send it a SIGTERM
```
strace: Process 3519 attached
write(29, "b3519\n", 6)                 = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGTERM {si_signo=SIGTERM, si_code=SI_USER, si_pid=5347, si_uid=0} ---
write(3, "\1\0\0\0\0\0\0\0", 8)         = 8
timer_settime(0, 0, {it_interval={0, 100000000}, it_value={0, 100000000}}, NULL) = 0
--- SIGVTALRM {si_signo=SIGVTALRM, si_code=SI_TKILL, si_pid=3519, si_uid=65534} ---
rt_sigreturn({mask=[TERM]})             = 0
--- SIGVTALRM {si_signo=SIGVTALRM, si_code=SI_TKILL, si_pid=3519, si_uid=65534} ---
rt_sigreturn({mask=[TERM]})             = 0
rt_sigreturn({mask=[]})                 = -1 EINTR (Interrupted system call)
timer_settime(0, 0, {it_interval={0, 0}, it_value={0, 0}}, NULL) = 0
write(43, "?", 1)                       = 1
write(29, "b3519\n", 6
^C
```

Non idle worker:

```
# These actually have a server loop, but also did not respond to SIGTERM
/home/app/lib/puma_helpers/hooks.rb:55:in `backtrace'
/home/app/lib/puma_helpers/hooks.rb:55:in `block (2 levels) in after_worker_fork'
/home/app/lib/puma_helpers/hooks.rb:47:in `each'
/home/app/lib/puma_helpers/hooks.rb:47:in `block in after_worker_fork'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:308:in `join'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:308:in `worker'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:139:in `block (2 levels) in spawn_workers'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:139:in `fork'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:139:in `block in spawn_workers'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:135:in `times'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:135:in `spawn_workers'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:213:in `check_workers'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cluster.rb:486:in `run'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/launcher.rb:186:in `run'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/lib/puma/cli.rb:80:in `run'
/home/app/vendor/bundle/ruby/2.6.0/gems/puma-3.12.2/bin/puma:10:in `<top (required)>'
/home/app/vendor/bundle/ruby/2.6.0/bin/puma:23:in `load'
/home/app/vendor/bundle/ruby/2.6.0/bin/puma:23:in `<top (required)>'
```

As you can see from the strace and the thread backtraces it looks like the main thread is blocked on reading/writing to the pipe and it's unable to do anything else. It may be possible to reproduce this by slowing down worker boot and running a large number of workers (we had 72).

**Comment by nateberkopec** [comment]
Closing as stale.

The signs here to me are pointing to something that will either be fixed _or_ completely changed by the nio reactor change in 4.0. If someone can reproduce this on 4.0 we can take it up.
